wallu/wallu.ino

```diff
@@ -0,0 +1,202 @@
+
+// HUD
+int battery_level = -1;
+unsigned int rpm_right = -1;
+unsigned int rpm_left = -1;
+byte lock_status = 0;    // 1 for locked, 0 for unlocked
+
+//Motor motor1 = Motor(9, 8, 7);
+
+void self_test() {
+  strcpy(com_buff, "MFM0");
+  com_buff_count = 4;
+  process_command();
+  Serial.println(requested_dir);
+  Serial.println(requested_throttle);
+}
+
+void process_movement(MotorDirection dir, int throttle, int angle) {
+  if (dir == BACKWARD) {
+    //Run normally
+  }
+  else { // if it's forward, we need to account for turning
+    
+  }
+}
+
+int calculate_battery(int read_in) {
+  // Vout = Vs x R2 / R1 + R2
+  // Vout * (R1 + R2) / R2 = Vs
+  float val = analogRead(read_in);
+  float input_voltage = val / 1023.0 * 5.0;
+  float raw_voltage = (input_voltage * (BATTERY_R1 + BATTERY_R2) / BATTERY_R2);
+  if (raw_voltage < BATT_MIN)
+    return 0;
+  return 100.0 * (raw_voltage - BATT_MIN) / (BATT_MAX - BATT_MIN);
+}
+
+void process_command() {
+  char atoi_buff[2];
+  if (com_buff_count == 1) {
+    ;
+  }
+  // Servo commands S_ _ _ (angle)
+  if (com_buff_count == 4) {
+    ;
+  }
+  // Motor Commands
+  if (com_buff_count == 6) {
+    if (com_buff[0] == 'M') {
+      if (com_buff[1] == 'F')
+        requested_dir = FORWARD;
+      else if (com_buff[1] == 'B')
+        requested_dir = BACKWARD;
+      else if (com_buff[1] == 'S')
+        requested_dir = STOP;
+      else
+        error_flag = 1; 
+    }
+    if (isDigit(com_buff[2]) && isDigit(com_buff[3])) {
+      atoi_buff[0] = com_buff[2];
+      atoi_buff[1] = com_buff[3];
+      requested_steering_angle = atoi(atoi_buff);
+    }
+    else
+      error_flag = 1;
+    if (com_buff[4] == 'M') //max speed
+      requested_throttle = 100;
+    else if (isDigit(com_buff[4]) && isDigit(com_buff[5]))
+      requested_throttle = atoi(com_buff+4);
+    else
+      error_flag = 1;
+  }
+}
+
+void check_for_command() {
+  //buffer incoming data until ; indicates end of command
+  char incomingByte;
+  while (Serial.available() > 0) {
+    incomingByte = Serial.read();
+    if (incomingByte == ';') {
+      process_command();
+      memset(com_buff, 0, sizeof(com_buff));
+      com_buff_count = 0;
+    }
+    else {
+      com_buff[com_buff_count] = incomingByte;
+      com_buff_count += 1;
+    }
+  }
+}
+ 
+void setup() {
+  // Initialization of all Arduino pins
+  pinMode(BATTERY_MONITOR_PIN, INPUT);
+  pinMode(LEFT_RPM_PIN, INPUT);
+  pinMode(RIGHT_RPM_PIN, INPUT);
+  pinMode(MOTOR_R_PWM, OUTPUT);
+  pinMode(MOTOR_R_IN1, OUTPUT);
+  pinMode(MOTOR_R_IN2, OUTPUT);
+  pinMode(MOTOR_R_PWM, OUTPUT);
+  pinMode(MOTOR_L_IN1, OUTPUT);
+  pinMode(MOTOR_L_IN2, OUTPUT);
+  pinMode(LOCK_SERVO_PIN, OUTPUT);
+  pinMode(CAMERA_SERVO_PIN, OUTPUT);
+  pinMode(LOCK_SENSOR_PIN, INPUT);
+  pinMode(RANGE_TRIG_PIN, INPUT); // need to double check
+  pinMode(RANGE_ECHO_PIN, INPUT);
+  pinMode(RED_LED_EN_PIN, OUTPUT);
+  pinMode(GREEN_LED_EN_PIN, OUTPUT);
+  pinMode(BLUE_LED_EN_PIN, OUTPUT);
```

Comment:
If we have a class for each of these devices, we can move the `pinMode()` calls to the constructor of each class. Reduces clutter here, since we want our main modules to be as high level as possible.